server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -1374,8 +1379,8 @@ private static class IndexMetadataDiff implements Diff<IndexMetadata> {
             aliases = DiffableUtils.diff(before.aliases, after.aliases, DiffableUtils.getStringKeySerializer());
             customData = DiffableUtils.diff(before.customData, after.customData, DiffableUtils.getStringKeySerializer());
             inSyncAllocationIds = DiffableUtils.diff(
-                before.inSyncAllocationIds,
-                after.inSyncAllocationIds,
+                Maps.fromList(before.inSyncAllocationIds),
+                Maps.fromList(after.inSyncAllocationIds),
```

Comment:
Hi @henningandersen , indeed it is unfortunate we need to keep supporting somehow the older map diff for bwc. I think this is still better than before because the crucial point of the ticket was to decrease memory, which I think is done since we keep an array now in the IndexMetadata (so we do not have the keys anymore in memory). The Map is only constructed when it needs to be streamed, for bwc purposes, so it is a bit more computation but hopefully does not impact the memory too much (since I hope that the streaming does not happen too often?).

I can make it version-dependent here to produce an "all" diff (basically passing the ArrayList as is) for future versions. But I am unaware of the proxy node problem. What do we mean by a "proxy" node here, do we mean a ES node in the cluster?

Reply:
Gentle reminder for more help/clarifications on this point.

Reply:
Gentle reminder for reviews and more help/clarifications on this point.

Reply:
A proxy node/action is for remote cluster communication, either CCS or CCR. A cluster only has so many connections to another cluster and may not even know the individual addresses on the other cluster (in case it goes through a proxy).

In case a local node wants to talk to a node in a remote cluster that it is not directly connected to, it will send the request to one of the nodes in that cluster. That node then acts as a proxy (for the specific request), forwarding the request to the right node. See also `TransportActionProxy`.

> The Map is only constructed when it needs to be streamed, for bwc purposes

Unfortunately, this is a forever bwc concern unless we do something now. I think we should take steps to remove this in this PR such that we can clear this out eventually. I wonder if there is anything around the versions, proxy action and diffing that can help us, I could imagine other problems going through a proxy action when diffing.

Reply:
I see, thanks for the expalantion! I do see in the incoming/outgoing messages that we can check the version, so I think we can indeed do what you had suggested:

> So in order to get there I think we need to stream both forms until next major and then after that, only stream the all-or-nothing form.

I will try to see that then, so that we stream the all-or-nothing form for current/future versions and the older diff map for older versions. And we can create a ticket to remove the diff map in the future.

Unless somebody has better idea(s) :)